initial sync, testnet, panic: attempt to write cfheaders out of order

lnd version: Version 0.5.0-beta commit=a3edcf9cd537ea37a502efc91d262e7017a99e6b
btcd version: 2a560b2036bee5e3679ec2133eb6520b2f195213

```
2018-10-08 11:06:19.826 [INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
2018-10-08 11:06:19.839 [INF] BTCN: Attempting to query for 1432 cfheader batches
2018-10-08 11:06:19.839 [DBG] BTCN: Sending getcfheaders to 207.154.224.115:18333 (outbound)
2018-10-08 11:06:22.841 [DBG] BTCN: Sending getcfheaders to 207.154.224.115:18333 (outbound)
2018-10-08 11:06:23.008 [DBG] BTCN: Received cfheaders from 207.154.224.115:18333 (outbound)
2018-10-08 11:06:23.010 [DBG] BTCN: Got cfheaders from height=1 to height=1001
2018-10-08 11:06:23.014 [DBG] BTCN: Writing filter headers up to height=1000
2018-10-08 11:06:23.019 [INF] BTCN: Verified 1 filter header in the last 4m52.34s (height 1, 2011-02-03 02:22:08 +0300 MSK)
2018-10-08 11:06:23.019 [DBG] BTCN: Sending getcfheaders to 207.154.224.115:18333 (outbound)
2018-10-08 11:06:23.056 [DBG] BTCN: Received cfheaders from 207.154.224.115:18333 (outbound)
2018-10-08 11:06:25.352 [DBG] BTCN: Received cfheaders from 207.154.224.115:18333 (outbound)
2018-10-08 11:06:25.354 [DBG] BTCN: Got cfheaders from height=1001 to height=2001
panic: couldn't write cfheaders msg: attempt to write cfheaders out of order

goroutine 54 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders.func1(0xc4200e0960, 0x1dbeea0, 0xc42048e3f0, 0x1dbec60, 0xc43102a900, 0x501)
	/Users/andrey/GoProjects/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:871 +0xb2d
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*ChainService).queryBatch(0xc4204a8000, 0xc431440000, 0x598, 0x598, 0xc430f0be30, 0xc420856000, 0x0, 0x0, 0x0)
	/Users/andrey/GoProjects/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/query.go:422 +0xa85
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders(0xc4202ee160, 0xc431430000, 0x598, 0x598, 0xc4202a2088, 0x500)
	/Users/andrey/GoProjects/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:774 +0x8d7
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler.func2(0xc42046a030, 0xc4202ee160, 0x15dc6b, 0xc4308c4840, 0x0, 0x1d146f8)
	/Users/andrey/GoProjects/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:500 +0x3c5
created by github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler
	/Users/andrey/GoProjects/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:440 +0x407
```

full log:
[lnd.log](https://github.com/lightninglabs/neutrino/files/2455127/neutrino.log)

I'm getting the same issue.
```
2018-10-08 14:15:43.433 [INF] LTND: LightningWallet opened
2018-10-08 14:15:44.613 [INF] BTCN: Fetching set of checkpointed cfheaders filters from height=60000, hash=9ac5e0eaa5360f49a74de72392ee7c1154eadbe4a7b1b836b997e6fcb94dce57
2018-10-08 14:15:44.613 [INF] BTCN: Starting to query for cfheaders from checkpoint_interval=60
2018-10-08 14:15:47.674 [INF] BTCN: Attempting to query for 24 cfheader batches
2018-10-08 14:15:47.919 [INF] LNWL: Started rescan from block 0000000008ca11392fa91c4786e59823a002f4868bdb0c1385b12a2844cbc11f (height 20000) for 0 addresses
2018-10-08 14:15:47.930 [INF] HSWC: Restoring in-memory circuit state from disk
2018-10-08 14:15:47.933 [INF] HSWC: Payment circuits loaded: num_pending=0, num_open=0
2018-10-08 14:15:47.943 [INF] RPCS: RPC server listening on [::]:10009
2018-10-08 14:15:47.943 [INF] LTND: Waiting for chain backend to finish sync, start_height=86546
2018-10-08 14:15:47.943 [INF] RPCS: gRPC proxy started at 127.0.0.1:8080
2018-10-08 14:15:53.553 [INF] BTCN: Verified 1001 filter headers in the last 12.82s (height 61001, 2013-03-20 05:45:28 +0000 GMT)
2018-10-08 14:15:54.053 [INF] BTCN: Processed 4001 blocks in the last 13.32s (height 88547, 2013-06-17 05:29:50 +0100 BST)
2018-10-08 14:16:04.270 [INF] BTCN: Verified 4000 filter headers in the last 10.71s (height 65001, 2013-04-06 14:38:09 +0100 BST)
panic: couldn't write cfheaders msg: attempt to write cfheaders out of order

goroutine 70 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders.func1(0x1960000, 0xc08dd8, 0x18a2c60, 0xc08cb8, 0x19680f0, 0x1)
	/home/videah/go/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:946 +0x8d4
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*ChainService).queryBatch(0x18acc40, 0x19ae580, 0x18, 0x54, 0x12fc5cb8, 0x1a3e580, 0x0, 0x0, 0x0)
	/home/videah/go/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/query.go:422 +0x7a8
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders(0x1902240, 0x11ec8000, 0x54, 0x80, 0x1a120f8, 0x0)
	/home/videah/go/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:849 +0x6c0
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler(0x1902240)
	/home/videah/go/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:560 +0xa88
created by github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).Start
	/home/videah/go/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:251 +0xe0
```

I have the same issue with running fullnode lnd with neutrino (local updated btcd).
Opeating System:  Windows 10.
```
2018-10-21 00:17:18.309 [DBG] BTCN: Got cfheaders from height=311001 to height=312001, prev_hash=83ef1920443a21a21cbc1dd43a8901e2b8710bab02e30e6a5cd828f2d661b934
2018-10-21 00:17:18.314 [DBG] BTCN: Writing filter headers up to height=312000, hash=00000000000000795751022d133ef4ed3cb395643f53816ed7c7494e910e5793, new_tip=10808d102ed466a116ca94b8c922774d518063e3c5f6d53455cf744a1ad01869
2018-10-21 00:17:21.229 [DBG] BTCN: Sending getcfheaders to 127.0.0.1:18333 (outbound)
2018-10-21 00:17:21.272 [DBG] BTCN: Sending getheaders (locator 00000000006f44fb3d464449c23c6815e0e457d2c0acc64a998a766bc489276c, stop 000000000000406178b12a4dea3b27e13b3c4fe4510994fd667d7c1e6a3f4dc1) to 127.0.0.1:18333 (outbound)
2018-10-21 00:17:21.312 [DBG] BTCN: Received cfheaders from 127.0.0.1:18333 (outbound)
2018-10-21 00:17:21.318 [DBG] BTCN: Received headers (num 2000) from 127.0.0.1:18333 (outbound)
2018-10-21 00:17:21.335 [DBG] BTCN: Difficulty retarget at block height 661248
2018-10-21 00:17:21.335 [DBG] BTCN: Old target 1c010f82 (00000000010f8200000000000000000000000000000000000000000000000000)
2018-10-21 00:17:21.337 [DBG] BTCN: New target 1b43e080 (000000000043e080000000000000000000000000000000000000000000000000)
2018-10-21 00:17:21.341 [DBG] BTCN: Actual timespan 1h34m58s, adjusted timespan 84h0m0s, target timespan 336h0m0s
2018-10-21 00:17:21.420 [DBG] BTCN: Wrote filter headers up to height=312000, hash=00000000000000795751022d133ef4ed3cb395643f53816ed7c7494e910e5793, new_tip=10808d102ed466a116ca94b8c922774d518063e3c5f6d53455cf744a1ad01869
2018-10-21 00:17:21.420 [INF] BTCN: Verified 6000 filter headers in the last 12.48s (height 311001, 2014-11-27 07:59:20 -0500 EST)
2018-10-21 00:17:21.422 [DBG] BTCN: Sending getcfheaders to 127.0.0.1:18333 (outbound)
2018-10-21 00:17:21.423 [DBG] BTCN: Got cfheaders from height=311001 to height=312001, prev_hash=83ef1920443a21a21cbc1dd43a8901e2b8710bab02e30e6a5cd828f2d661b934
panic: couldn't write cfheaders msg: attempt to write cfheaders out of order! Tip=10808d102ed466a116ca94b8c922774d518063e3c5f6d53455cf744a1ad01869 (height=312000), prev_hash=83ef1920443a21a21cbc1dd43a8901e2b8710bab02e30e6a5cd828f2d661b934.

goroutine 63 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders.func1(0xc00024c640, 0x122fb00, 0xc0004c37a0, 0x122f8c0, 0xc012058300, 0x1)
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:945 +0xbff
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.queryChainServiceBatch(0xc000014dc0, 0xc01343a000, 0xd4, 0x180, 0xc0111ef560, 0xc00005ecc0, 0x0, 0x0, 0x0)
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/query.go:425 +0xa24
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.NewChainService.func3(0xc01343a000, 0xd4, 0x180, 0xc0111ef560, 0xc00005ecc0, 0x0, 0x0, 0x0)
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/neutrino.go:624 +0x8a
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders(0xc000284420, 0xc0114ba000, 0x180, 0x200, 0xc0004760d8, 0x100)
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:843 +0x7d9
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler(0xc000284420)
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:559 +0xc9b
created by github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).Start
        C:/Users/moli/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:251 +0x109
```
Full log:  https://dpaste.de/QKwV/raw

@halseth per your request, with #105 and a patch to include line 897 change from #106, using simnet I get this:
```
2018-10-21 23:46:30.580 [INF] LTND: Version: 0.5.0-beta commit=v0.4.2-beta-1247-ge22ae9985b696924d5934004c269c930d3ccba43-dirty, build=production, logging=default
2018-10-21 23:46:30.580 [INF] LTND: Active chain: Bitcoin (network=simnet)
2018-10-21 23:46:30.581 [INF] CHDB: Checking for schema update: latest_version=6, db_version=6
2018-10-21 23:46:30.597 [INF] LTND: Primary chain is set to: bitcoin
2018-10-21 23:46:30.601 [INF] BTCN: Loaded 0 addresses from file 'data/chain/bitcoin/simnet/peers.json'
2018-10-21 23:46:30.601 [INF] BTCN: Waiting for more block headers, then will start cfheaders sync from height 0...
2018-10-21 23:46:30.601 [DBG] CMGR: Attempting to connect to 35.231.222.142:18555 (reqid 1)
2018-10-21 23:46:30.779 [DBG] CMGR: Connected to 35.231.222.142:18555 (reqid 1)
2018-10-21 23:46:30.779 [DBG] BTCN: Sending version (agent /btcwire:0.5.0/neutrino:0.0.4-beta/, pver 70013, block 0) to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.031 [DBG] BTCN: Received version (agent /btcwire:0.5.0/btcd:0.12.0/, pver 70013, block 12667) from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.031 [DBG] BTCN: Negotiated protocol version 70013 for peer 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.031 [DBG] BTCN: Added time sample of 0s (total: 1)
2018-10-21 23:46:31.031 [DBG] BTCN: Connected to 35.231.222.142:18555
2018-10-21 23:46:31.031 [DBG] BTCN: New peer 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.031 [INF] BTCN: New valid peer 35.231.222.142:18555 (outbound) (/btcwire:0.5.0/btcd:0.12.0/)
2018-10-21 23:46:31.031 [INF] BTCN: Syncing to block height 12667 from peer 35.231.222.142:18555
2018-10-21 23:46:31.031 [INF] BTCN: Fetching set of headers from tip (height=0) from peer 35.231.222.142:18555
2018-10-21 23:46:31.031 [DBG] BTCN: Sending verack to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.031 [DBG] BTCN: Sending getheaders (locator 683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:31.041 [DBG] BTCN: Received verack from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:33.285 [INF] LNWL: Opened wallet
2018-10-21 23:46:34.165 [INF] LNWL: The wallet has been unlocked without a time limit
2018-10-21 23:46:34.165 [INF] LNWL: Catching up block hashes to height 0, this will take a while...
2018-10-21 23:46:34.166 [INF] LTND: LightningWallet opened
2018-10-21 23:46:35.893 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:35.925 [INF] BTCN: Starting cfheaders sync from (block_height=0, block_hash=683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6) to (block_height=2000, block_hash=6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3)
2018-10-21 23:46:35.925 [DBG] BTCN: Sending getheaders (locator 6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:35.925 [INF] BTCN: Starting cfheaders sync for filter_type=0
2018-10-21 23:46:35.925 [DBG] BTCN: Getting filter checkpoints up to height=2000, hash=6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3
2018-10-21 23:46:35.925 [DBG] BTCN: Sending getcfcheckpt to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:37.960 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:37.961 [DBG] BTCN: Received cfcheckpt from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:37.961 [DBG] BTCN: Received unhandled message of type cfcheckpt from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:37.961 [INF] BTCN: Fetching set of checkpointed cfheaders filters from height=0, hash=cf24d6ca4862839dbafe633c144fd92df0ceb3a7c9c99fd73208795d5c0fefb2
2018-10-21 23:46:37.961 [INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
2018-10-21 23:46:37.961 [INF] BTCN: Attempting to query for 2 cfheader batches
2018-10-21 23:46:37.961 [DBG] BTCN: Sending getcfheaders to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:37.961 [DBG] BTCN: Difficulty retarget at block height 2016
2018-10-21 23:46:37.961 [DBG] BTCN: Old target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-21 23:46:37.961 [DBG] BTCN: New target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-21 23:46:37.961 [DBG] BTCN: Actual timespan 38239h30m45s, adjusted timespan 1344h0m0s, target timespan 336h0m0s
2018-10-21 23:46:37.997 [DBG] BTCN: Sending getheaders (locator 255afa63c71207662247a1d97edb1c98146e891aa0fcbda2484e78e627cfaf24, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:39.114 [DBG] BTCN: Received cfheaders from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:39.116 [DBG] BTCN: Got cfheaders from height=1 to height=1000, prev_hash=4d4cb682e2416c4dfca849863be8fe2da0dad1010bc3ad67f0687882c37d067a
2018-10-21 23:46:39.116 [DBG] BTCN: Using offset 0 for initial filter header range (new prev_hash=cf24d6ca4862839dbafe633c144fd92df0ceb3a7c9c99fd73208795d5c0fefb2)
2018-10-21 23:46:39.119 [DBG] BTCN: Writing filter headers up to height=1000, hash=773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1
2018-10-21 23:46:39.121 [DBG] BTCN: Sending getcfheaders to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:39.143 [INF] LNWL: Caught up to height 0
2018-10-21 23:46:39.146 [INF] HSWC: Restoring in-memory circuit state from disk
2018-10-21 23:46:39.147 [INF] HSWC: Payment circuits loaded: num_pending=0, num_open=0
2018-10-21 23:46:39.148 [INF] RPCS: RPC server listening on 127.0.0.1:10002
2018-10-21 23:46:39.148 [INF] RPCS: gRPC proxy started at 127.0.0.1:8002
2018-10-21 23:46:39.149 [DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1000)
2018-10-21 23:46:39.149 [DBG] BTCN: Starting rescan from known block 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1)
2018-10-21 23:46:39.149 [DBG] BTCN: Rescan became current at 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1), subscribing to block notifications
2018-10-21 23:46:39.149 [INF] HSWC: Starting HTLC Switch
2018-10-21 23:46:39.149 [INF] NTFN: New block epoch subscription
2018-10-21 23:46:39.149 [INF] NTFN: New block epoch subscription
2018-10-21 23:46:39.150 [INF] NTFN: New block epoch subscription
2018-10-21 23:46:39.150 [INF] DISC: Authenticated Gossiper is starting
2018-10-21 23:46:39.150 [INF] BRAR: Starting contract observer, watching for breaches.
2018-10-21 23:46:39.150 [INF] NTFN: New block epoch subscription
2018-10-21 23:46:39.150 [INF] CRTR: FilteredChainView starting
2018-10-21 23:46:39.150 [DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1000)
2018-10-21 23:46:39.150 [DBG] BTCN: Starting rescan from known block 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1)
2018-10-21 23:46:39.150 [DBG] BTCN: Rescan became current at 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1), subscribing to block notifications
2018-10-21 23:46:39.151 [INF] CRTR: Filtering chain using 0 channels active
2018-10-21 23:46:39.151 [INF] CRTR: Prune tip for Channel Graph: height=1000, hash=773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1
2018-10-21 23:46:39.151 [INF] CMGR: Server listening on [::]:10012
2018-10-21 23:46:39.151 [INF] SRVR: Auto peer bootstrapping is disabled
2018-10-21 23:46:42.122 [DBG] BTCN: Sending getcfheaders to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:43.021 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:43.021 [INF] BTCN: Processed 4001 blocks in the last 12.42s (height 4001, 2018-10-09 11:16:39 +0300 IDT)
2018-10-21 23:46:43.022 [DBG] BTCN: Difficulty retarget at block height 4032
2018-10-21 23:46:43.022 [DBG] BTCN: Old target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-21 23:46:43.022 [DBG] BTCN: New target 201fffff (1fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-21 23:46:43.022 [DBG] BTCN: Actual timespan 33h23m4s, adjusted timespan 84h0m0s, target timespan 336h0m0s
2018-10-21 23:46:43.042 [DBG] BTCN: Sending getheaders (locator 1e7f8e79706074a692e8e0b94320d1000ef643bd096ba261bf4f203a37788713, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-21 23:46:43.475 [DBG] BTCN: Received cfheaders from 35.231.222.142:18555 (outbound)
2018-10-21 23:46:43.477 [DBG] BTCN: Got cfheaders from height=1001 to height=2000, prev_hash=781069242de4f4360e0e05ec8e79fa02c5a5daec5dacdfc88e075e3ec7830837
panic: couldn't write cfheaders msg: attempt to write cfheaders out of order! Tip=7b033e39f2ba21c0b92ab2a3f75cf5fa0a9ebf093f9859ad1085610bbe9c67ab, prev_hash=781069242de4f4360e0e05ec8e79fa02c5a5daec5dacdfc88e075e3ec7830837.

goroutine 26 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders.func1(0xc42049c0a0, 0x1dd7d00, 0xc42003a420, 0x1dd7ac0, 0xc42009e060, 0x1)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:945 +0xc4e
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.queryChainServiceBatch(0xc4204ec840, 0xc4308fec60, 0x2, 0x2, 0xc4207bc960, 0xc4200aa900, 0x0, 0x0, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/query.go:425 +0xa85
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.NewChainService.func3(0xc4308fec60, 0x2, 0x2, 0xc4207bc960, 0xc4200aa900, 0x0, 0x0, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/neutrino.go:624 +0x83
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders(0xc4202622c0, 0xc4200b5e80, 0x2, 0x2, 0xc42022e0a8, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:843 +0x878
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler(0xc4202622c0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:559 +0xda6
created by github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).Start
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:251 +0x107
admins-MacBook-Pro-2:btc admin$ 
```

```
2018-10-22 00:11:04.299 [INF] LTND: Version: 0.5.0-beta commit=, build=production, logging=default
2018-10-22 00:11:04.299 [INF] LTND: Active chain: Bitcoin (network=simnet)
2018-10-22 00:11:04.300 [INF] CHDB: Checking for schema update: latest_version=6, db_version=6
2018-10-22 00:11:04.317 [INF] LTND: Primary chain is set to: bitcoin
2018-10-22 00:11:04.322 [INF] BTCN: Loaded 0 addresses from file 'data/chain/bitcoin/simnet/peers.json'
2018-10-22 00:11:04.322 [INF] BTCN: Waiting for more block headers, then will start cfheaders sync from height 0...
2018-10-22 00:11:04.322 [DBG] CMGR: Attempting to connect to 35.231.222.142:18555 (reqid 1)
2018-10-22 00:11:04.490 [DBG] CMGR: Connected to 35.231.222.142:18555 (reqid 1)
2018-10-22 00:11:04.490 [DBG] BTCN: Sending version (agent /btcwire:0.5.0/neutrino:0.0.4-beta/, pver 70013, block 0) to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [DBG] BTCN: Received version (agent /btcwire:0.5.0/btcd:0.12.0/, pver 70013, block 12690) from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [DBG] BTCN: Negotiated protocol version 70013 for peer 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [DBG] BTCN: Added time sample of 1s (total: 1)
2018-10-22 00:11:04.660 [DBG] BTCN: Connected to 35.231.222.142:18555
2018-10-22 00:11:04.660 [DBG] BTCN: New peer 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [DBG] BTCN: Received verack from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [INF] BTCN: New valid peer 35.231.222.142:18555 (outbound) (/btcwire:0.5.0/btcd:0.12.0/)
2018-10-22 00:11:04.660 [INF] BTCN: Syncing to block height 12690 from peer 35.231.222.142:18555
2018-10-22 00:11:04.660 [INF] BTCN: Fetching set of headers from tip (height=0) from peer 35.231.222.142:18555
2018-10-22 00:11:04.660 [DBG] BTCN: Sending verack to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:04.660 [DBG] BTCN: Sending getheaders (locator 683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:06.434 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:06.465 [INF] BTCN: Starting cfheaders sync from (block_height=0, block_hash=683e86bd5c6d110d91b94b97137ba6bfe02dbbdb8e3dff722a669b5d69d77af6) to (block_height=2000, block_hash=6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3)
2018-10-22 00:11:06.465 [INF] BTCN: Starting cfheaders sync for filter_type=0
2018-10-22 00:11:06.465 [DBG] BTCN: Getting filter checkpoints up to height=2000, hash=6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3
2018-10-22 00:11:06.465 [DBG] BTCN: Sending getheaders (locator 6d11267dc384cf0bf0d8e137e13a63c9606a54379e104a8caa57001f14d821f3, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:06.465 [DBG] BTCN: Sending getcfcheckpt to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.021 [INF] LNWL: Opened wallet
2018-10-22 00:11:07.651 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.651 [DBG] BTCN: Received cfcheckpt from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.651 [DBG] BTCN: Received unhandled message of type cfcheckpt from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.651 [INF] BTCN: Fetching set of checkpointed cfheaders filters from height=0, hash=cf24d6ca4862839dbafe633c144fd92df0ceb3a7c9c99fd73208795d5c0fefb2
2018-10-22 00:11:07.651 [INF] BTCN: Starting to query for cfheaders from checkpoint_interval=0
2018-10-22 00:11:07.651 [INF] BTCN: Attempting to query for 2 cfheader batches
2018-10-22 00:11:07.651 [DBG] BTCN: Difficulty retarget at block height 2016
2018-10-22 00:11:07.651 [DBG] BTCN: Old target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-22 00:11:07.651 [DBG] BTCN: Sending getcfheaders to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.651 [DBG] BTCN: New target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-22 00:11:07.651 [DBG] BTCN: Actual timespan 38239h30m45s, adjusted timespan 1344h0m0s, target timespan 336h0m0s
2018-10-22 00:11:07.683 [DBG] BTCN: Sending getheaders (locator 255afa63c71207662247a1d97edb1c98146e891aa0fcbda2484e78e627cfaf24, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:07.886 [INF] LNWL: The wallet has been unlocked without a time limit
2018-10-22 00:11:07.886 [INF] LNWL: Catching up block hashes to height 0, this will take a while...
2018-10-22 00:11:07.888 [INF] LTND: LightningWallet opened
2018-10-22 00:11:08.020 [DBG] BTCN: Received cfheaders from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:08.022 [DBG] BTCN: Got cfheaders from height=1 to height=1000, prev_hash=4d4cb682e2416c4dfca849863be8fe2da0dad1010bc3ad67f0687882c37d067a
2018-10-22 00:11:08.022 [DBG] BTCN: Using offset 0 for initial filter header range (new prev_hash=cf24d6ca4862839dbafe633c144fd92df0ceb3a7c9c99fd73208795d5c0fefb2)
2018-10-22 00:11:08.024 [DBG] BTCN: Writing filter headers up to height=1000, hash=773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1, new_tip=7b033e39f2ba21c0b92ab2a3f75cf5fa0a9ebf093f9859ad1085610bbe9c67ab
2018-10-22 00:11:08.026 [DBG] BTCN: Wrote filter headers up to height=1000, hash=773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1, new_tip=7b033e39f2ba21c0b92ab2a3f75cf5fa0a9ebf093f9859ad1085610bbe9c67ab
2018-10-22 00:11:08.026 [DBG] BTCN: Sending getcfheaders to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:08.097 [INF] LNWL: Caught up to height 0
2018-10-22 00:11:08.100 [INF] HSWC: Restoring in-memory circuit state from disk
2018-10-22 00:11:08.101 [INF] HSWC: Payment circuits loaded: num_pending=0, num_open=0
2018-10-22 00:11:08.102 [INF] RPCS: RPC server listening on 127.0.0.1:10002
2018-10-22 00:11:08.102 [INF] RPCS: gRPC proxy started at 127.0.0.1:8002
2018-10-22 00:11:08.102 [DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1000)
2018-10-22 00:11:08.102 [DBG] BTCN: Starting rescan from known block 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1)
2018-10-22 00:11:08.102 [DBG] BTCN: Rescan became current at 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1), subscribing to block notifications
2018-10-22 00:11:08.103 [INF] HSWC: Starting HTLC Switch
2018-10-22 00:11:08.103 [INF] NTFN: New block epoch subscription
2018-10-22 00:11:08.103 [INF] NTFN: New block epoch subscription
2018-10-22 00:11:08.103 [INF] NTFN: New block epoch subscription
2018-10-22 00:11:08.103 [INF] DISC: Authenticated Gossiper is starting
2018-10-22 00:11:08.103 [INF] BRAR: Starting contract observer, watching for breaches.
2018-10-22 00:11:08.103 [INF] NTFN: New block epoch subscription
2018-10-22 00:11:08.103 [INF] CRTR: FilteredChainView starting
2018-10-22 00:11:08.103 [DBG] BTCN: Waiting for filter headers (height=1000) to catch up the rescan start (height=1000)
2018-10-22 00:11:08.103 [DBG] BTCN: Starting rescan from known block 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1)
2018-10-22 00:11:08.104 [DBG] BTCN: Rescan became current at 1000 (773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1), subscribing to block notifications
2018-10-22 00:11:08.105 [INF] CRTR: Filtering chain using 0 channels active
2018-10-22 00:11:08.105 [INF] CRTR: Prune tip for Channel Graph: height=1000, hash=773e6fc412cd790f4eac5ef4f67ddba99c3205ff0d69ec111bb6985476e692e1
2018-10-22 00:11:08.105 [INF] CMGR: Server listening on [::]:10012
2018-10-22 00:11:08.106 [INF] SRVR: Auto peer bootstrapping is disabled
2018-10-22 00:11:08.694 [DBG] BTCN: Received headers (num 2000) from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:08.695 [DBG] BTCN: Difficulty retarget at block height 4032
2018-10-22 00:11:08.695 [DBG] BTCN: Old target 207fffff (7fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-22 00:11:08.695 [DBG] BTCN: New target 201fffff (1fffff0000000000000000000000000000000000000000000000000000000000)
2018-10-22 00:11:08.695 [DBG] BTCN: Actual timespan 33h23m4s, adjusted timespan 84h0m0s, target timespan 336h0m0s
2018-10-22 00:11:08.720 [DBG] BTCN: Sending getheaders (locator 1e7f8e79706074a692e8e0b94320d1000ef643bd096ba261bf4f203a37788713, stop 0000000000000000000000000000000000000000000000000000000000000000) to 35.231.222.142:18555 (outbound)
2018-10-22 00:11:08.852 [DBG] BTCN: Received cfheaders from 35.231.222.142:18555 (outbound)
2018-10-22 00:11:08.854 [DBG] BTCN: Got cfheaders from height=1001 to height=2000, prev_hash=781069242de4f4360e0e05ec8e79fa02c5a5daec5dacdfc88e075e3ec7830837
panic: couldn't write cfheaders msg: attempt to write cfheaders out of order! Tip=7b033e39f2ba21c0b92ab2a3f75cf5fa0a9ebf093f9859ad1085610bbe9c67ab (height=1000), prev_hash=781069242de4f4360e0e05ec8e79fa02c5a5daec5dacdfc88e075e3ec7830837.

goroutine 52 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders.func1(0xc4204a20a0, 0x1dd45c0, 0xc4200d8630, 0x1dd4380, 0xc4204c6a80, 0x1)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:945 +0xc4e
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.queryChainServiceBatch(0xc4204f58c0, 0xc4201cefe0, 0x2, 0x2, 0xc4202524e0, 0xc42023a960, 0x0, 0x0, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/query.go:425 +0xa85
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.NewChainService.func3(0xc4201cefe0, 0x2, 0x2, 0xc4202524e0, 0xc42023a960, 0x0, 0x0, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/neutrino.go:624 +0x83
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).getCheckpointedCFHeaders(0xc42034cb00, 0xc4200b4750, 0x2, 0x2, 0xc42033c280, 0x0)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:843 +0x878
github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).cfHandler(0xc42034cb00)
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:559 +0xda6
created by github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino.(*blockManager).Start
	/Users/admin/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightninglabs/neutrino/blockmanager.go:251 +0x107
admins-MacBook-Pro-2:btc admin$ 
```

--- a/blockmanager.go
+++ b/blockmanager.go
@@ -893,7 +893,7 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
 			// Find the first and last height for the blocks
 			// represented by this message.
 			startHeight := checkPointIndex*wire.CFCheckptInterval + 1
-			lastHeight := startHeight + wire.CFCheckptInterval
+			lastHeight := (checkPointIndex + 1) * wire.CFCheckptInterval
 
 			log.Debugf("Got cfheaders from height=%v to "+
 				"height=%v, prev_hash=%v", startHeight,

--- a/blockmanager.go
+++ b/blockmanager.go
@@ -103,6 +103,9 @@ type blockManager struct {
 	// headers that we've verified in the past 10 seconds.
 	fltrHeaderProgessLogger *headerProgressLogger
 
+	// genesisHeader is the filter header of the genesis block.
+	genesisHeader chainhash.Hash
+
 	// headerTip will be set to the current block header tip at all times.
 	// Callers MUST hold the lock below each time they read/write from
 	// this field.
@@ -206,6 +209,14 @@ func newBlockManager(s *ChainService) (*blockManager, error) {
 	// duties.
 	bm.newHeadersSignal = sync.NewCond(&bm.newHeadersMtx)
 	bm.newFilterHeadersSignal = sync.NewCond(&bm.newFilterHeadersMtx)
+
+	// We fetch the genesis header to use for verifying the first received
+	// interval.
+	genesisHeader, err := s.RegFilterHeaders.FetchHeaderByHeight(0)
+	if err != nil {
+		return nil, err
+	}
+	bm.genesisHeader = *genesisHeader
 
 	// Initialize the next checkpoint based on the current height.
 	header, height, err := s.BlockHeaders.ChainTip()
@@ -874,8 +885,19 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
 				return false
 			}
 
+			// Use either the genesis header or the previous
+			// checkpoint index as the previous checkpoint when
+			// verifying that the filter headers in the response
+			// match up.
+			prevCheckpoint := &b.genesisHeader
+			if checkPointIndex > 0 {
+				prevCheckpoint = checkpoints[checkPointIndex-1]
+
+			}
+			nextCheckpoint := checkpoints[checkPointIndex]
+
 			// The response doesn't match the checkpoint.
-			if !verifyCheckpoint(checkpoints[checkPointIndex], r) {
+			if !verifyCheckpoint(prevCheckpoint, nextCheckpoint, r) {
 				log.Warnf("Checkpoints at index %v don't match " +
 					"response!!!")
 				return false
@@ -1111,9 +1133,14 @@ func minCheckpointHeight(checkpoints map[string][]*chainhash.Hash) uint32 {
 }
 
 // verifyHeaderCheckpoint verifies that a CFHeaders message matches the passed
-// checkpoint. It assumes everything else has been checked, including filter
+// checkpoints. It assumes everything else has been checked, including filter
 // type and stop hash matches, and returns true if matching and false if not.
-func verifyCheckpoint(checkpoint *chainhash.Hash, cfheaders *wire.MsgCFHeaders) bool {
+func verifyCheckpoint(prevCheckpoint, nextCheckpoint *chainhash.Hash,
+	cfheaders *wire.MsgCFHeaders) bool {
+
+	if *prevCheckpoint != cfheaders.PrevFilterHeader {
+		return false
+	}
 
 	lastHeader := cfheaders.PrevFilterHeader
 	for _, hash := range cfheaders.FilterHashes {
@@ -1122,7 +1149,7 @@ func verifyCheckpoint(checkpoint *chainhash.Hash, cfheaders *wire.MsgCFHeaders) 
 		)
 	}
 
-	return lastHeader == *checkpoint
+	return lastHeader == *nextCheckpoint
 }
 
 // resolveConflict finds the correct checkpoint information, rewinds the header

--- a/blockmanager.go
+++ b/blockmanager.go
@@ -1003,14 +1003,14 @@ func (b *blockManager) writeCFHeadersMsg(msg *wire.MsgCFHeaders,
 
 	// Check that the PrevFilterHeader is the same as the last stored so we
 	// can prevent misalignment.
-	tip, _, err := store.ChainTip()
+	tip, tipHeight, err := store.ChainTip()
 	if err != nil {
 		return nil, err
 	}
 	if *tip != msg.PrevFilterHeader {
 		return nil, fmt.Errorf("attempt to write cfheaders out of "+
-			"order! Tip=%v, prev_hash=%v.", *tip,
-			msg.PrevFilterHeader)
+			"order! Tip=%v (height=%v), prev_hash=%v.", *tip,
+			tipHeight, msg.PrevFilterHeader)
 	}
 
 	// Cycle through the headers and compute each header based on the prev
@@ -1054,8 +1054,8 @@ func (b *blockManager) writeCFHeadersMsg(msg *wire.MsgCFHeaders,
 	headerBatch[numHeaders-1].HeaderHash = lastHash
 	headerBatch[numHeaders-1].Height = lastHeight
 
-	log.Debugf("Writing filter headers up to height=%v, hash=%v",
-		lastHeight, lastHash)
+	log.Debugf("Writing filter headers up to height=%v, hash=%v, "+
+		"new_tip=%v", lastHeight, lastHash, lastHeader)
 
 	// Write the header batch.
 	err = store.WriteHeaders(headerBatch...)

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -22,7 +22,7 @@ import (
 func TestBlockManagerInitialInterval(t *testing.T) {
 	t.Parallel()
 
-	testCases := []struct {
+	type testCase struct {
 		// permute indicates whether responses should be permutated.
 		permute bool
 
@@ -30,23 +30,24 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 		// the first checkpoint interval to the filter header store
 		// before starting the test.
 		partialInterval bool
-	}{
-		{
-			permute:         false,
-			partialInterval: false,
-		},
-		{
-			permute:         false,
-			partialInterval: true,
-		},
-		{
-			permute:         true,
-			partialInterval: false,
-		},
-		{
-			permute:         true,
-			partialInterval: true,
-		},
+
+		// repeat indicates whether responses should be repeated.
+		repeat bool
+	}
+
+	// Generate all combinations of testcases.
+	var testCases []testCase
+	b := []bool{false, true}
+	for _, perm := range b {
+		for _, part := range b {
+			for _, rep := range b {
+				testCases = append(testCases, testCase{
+					permute:         perm,
+					partialInterval: part,
+					repeat:          rep,
+				})
+			}
+		}
 	}
 
 	for _, test := range testCases {
@@ -263,6 +264,18 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 				if !f(nil, msgs[index], responses[index]) {
 					t.Fatalf("got response false")
 				}
+
+				// If we are not testing repeated responses, go
+				// on to the next response.
+				if !test.repeat {
+					continue
+				}
+
+				// Otherwise resend the response we just sent.
+				if !f(nil, msgs[index], responses[index]) {
+					t.Fatalf("got response false")
+				}
+
 			}
 		}
 

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -415,6 +415,10 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 		// should not line up with the previous checkpoint.
 		intervalMisaligned bool
 
+		// invalidPrevHash indicates whether the interval responses
+		// should have a prev hash that doesn't mathc that interval.
+		invalidPrevHash bool
+
 		// partialInterval indicates whether we should write parts of
 		// the first checkpoint interval to the filter header store
 		// before starting the test.
@@ -456,6 +460,13 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 			intervalMisaligned: true,
 			partialInterval:    true,
 			firstInvalid:       1,
+		},
+
+		// With responses having invalid prev hashes, the second
+		// interval should be deemed invalid.
+		{
+			invalidPrevHash: true,
+			firstInvalid:    1,
 		},
 	}
 
@@ -539,6 +550,18 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 						continue
 					}
 					responses[i].PrevFilterHeader[0] ^= 1
+				}
+			}
+
+			// If we are testing for intervals with invalid prev
+			// hashes, we flip a bit to corrup them, regardless of
+			// whether we are testing misaligned intervals.
+			if test.invalidPrevHash {
+				for i := range responses {
+					if i == 0 {
+						continue
+					}
+					responses[i].PrevFilterHeader[1] ^= 1
 				}
 			}
 

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -400,3 +400,129 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 		}
 	}
 }
+
+// TestBlockManagerInvalidInterval tests that the block manager is able to
+// determine it is receiving corrupt checkpoints and filter headers.
+func TestBlockManagerInvalidInterval(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		// wrongGenesis indicates whether we should start deriving the
+		// filters from a wrong genesis.
+		wrongGenesis bool
+
+		// partialInterval indicates whether we should write parts of
+		// the first checkpoint interval to the filter header store
+		// before starting the test.
+		partialInterval bool
+
+		// firstInvalid is the first interval we expect the
+		// blockmanager to determine is invalid.
+		firstInvalid int
+	}
+
+	testCases := []testCase{
+		// With a set of checkpoints and filter headers calculated from
+		// the wrong genesis, the block manager should be able to
+		// determine that the first interval doesn't line up.
+		{
+			wrongGenesis: true,
+			firstInvalid: 0,
+		},
+
+		// With checkpoints calculated from the wrong genesis, and a
+		// partial set of filter headers already written, the first
+		// interval should be considered invalid.
+		{
+			wrongGenesis:    true,
+			partialInterval: true,
+			firstInvalid:    0,
+		},
+	}
+
+	for _, test := range testCases {
+		bm, hdrStore, cfStore, cleanUp, err := setupBlockManager()
+		if err != nil {
+			t.Fatalf("unable to set up ChainService: %v", err)
+		}
+		defer cleanUp()
+
+		// Keep track of the filter headers and block headers. Since
+		// the genesis headers are written automatically when the store
+		// is created, we query it to add to the slices.
+		genesisBlockHeader, _, err := hdrStore.ChainTip()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		genesisFilterHeader, _, err := cfStore.ChainTip()
+		if err != nil {
+			t.Fatal(err)
+		}
+		// To emulate a full node serving us filter headers derived
+		// from different genesis than what we have, we flip a bit in
+		// the genesis filter header.
+		if test.wrongGenesis {
+			genesisFilterHeader[0] ^= 1
+		}
+
+		headers, err := generateHeaders(genesisBlockHeader,
+			genesisFilterHeader, nil)
+		if err != nil {
+			t.Fatalf("unable to generate headers: %v", err)
+		}
+
+		// Write all block headers but the genesis, since it is already
+		// in the store.
+		if err = hdrStore.WriteHeaders(headers.blockHeaders[1:]...); err != nil {
+			t.Fatalf("Error writing batch of headers: %s", err)
+		}
+
+		// We emulate the case where a few filter headers are already
+		// written to the store by writing 1/3 of the first interval.
+		if test.partialInterval {
+			err = cfStore.WriteHeaders(
+				headers.cfHeaders[1 : wire.CFCheckptInterval/3]...,
+			)
+			if err != nil {
+				t.Fatalf("Error writing batch of headers: %s",
+					err)
+			}
+		}
+
+		bm.server.queryBatch = func(msgs []wire.Message,
+			f func(*ServerPeer, wire.Message, wire.Message) bool,
+			q <-chan struct{}, qo ...QueryOption) {
+
+			responses, err := generateResponses(msgs, headers)
+			if err != nil {
+				t.Fatalf("unable to generate responses: %v",
+					err)
+			}
+
+			// Check that the success of the callback match what we
+			// expect.
+			for i := range responses {
+				success := f(nil, msgs[i], responses[i])
+				if i == test.firstInvalid {
+					if success {
+						t.Fatalf("expected interval "+
+							"%d to be invalid", i)
+					}
+					break
+				}
+
+				if !success {
+					t.Fatalf("expected interval %d to be "+
+						"valid", i)
+				}
+			}
+		}
+
+		// Start the test by calling the get checkpointed cf headers
+		// method with the checkpoints we created.
+		bm.getCheckpointedCFHeaders(
+			headers.checkpoints, cfStore, wire.GCSFilterRegular,
+		)
+	}
+}

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -2,10 +2,10 @@ package neutrino
 
 import (
 	"encoding/binary"
+	"fmt"
 	"io/ioutil"
 	"math/rand"
 	"os"
-	"sync"
 	"testing"
 
 	"github.com/btcsuite/btcd/chaincfg"
@@ -15,6 +15,235 @@ import (
 	"github.com/btcsuite/btcwallet/walletdb"
 	"github.com/lightninglabs/neutrino/headerfs"
 )
+
+// maxHeight is the height we will generate filter headers up to.
+const maxHeight = 20 * uint32(wire.CFCheckptInterval)
+
+// setupBlockManager initialises a blockManager to be used in tests.
+func setupBlockManager() (*blockManager, headerfs.BlockHeaderStore,
+	*headerfs.FilterHeaderStore, func(), error) {
+
+	// Set up the block and filter header stores.
+	tempDir, err := ioutil.TempDir("", "neutrino")
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("Failed to create "+
+			"temporary directory: %s", err)
+	}
+
+	db, err := walletdb.Create("bdb", tempDir+"/weks.db")
+	if err != nil {
+		os.RemoveAll(tempDir)
+		return nil, nil, nil, nil, fmt.Errorf("Error opening DB: %s",
+			err)
+	}
+
+	hdrStore, err := headerfs.NewBlockHeaderStore(
+		tempDir, db, &chaincfg.SimNetParams,
+	)
+	if err != nil {
+		db.Close()
+		return nil, nil, nil, nil, fmt.Errorf("Error creating block "+
+			"header store: %s", err)
+	}
+
+	cleanUp := func() {
+		defer os.RemoveAll(tempDir)
+		defer db.Close()
+	}
+
+	cfStore, err := headerfs.NewFilterHeaderStore(
+		tempDir, db, headerfs.RegularFilter,
+		&chaincfg.SimNetParams,
+	)
+	if err != nil {
+		cleanUp()
+		return nil, nil, nil, nil, fmt.Errorf("Error creating filter "+
+			"header store: %s", err)
+	}
+
+	// Set up a chain service for the block manager. Each test should set
+	// custom query methods on this chain service.
+	cs := &ChainService{
+		chainParams:      chaincfg.SimNetParams,
+		BlockHeaders:     hdrStore,
+		RegFilterHeaders: cfStore,
+	}
+
+	// Set up a blockManager with the chain service we defined.
+	bm, err := newBlockManager(cs)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("unable to create "+
+			"blockmanager: %v", err)
+	}
+
+	return bm, hdrStore, cfStore, cleanUp, nil
+}
+
+// headers wraps the different headers and filters used throughout the tests.
+type headers struct {
+	blockHeaders []headerfs.BlockHeader
+	cfHeaders    []headerfs.FilterHeader
+	checkpoints  []*chainhash.Hash
+	filterHashes []chainhash.Hash
+}
+
+// generateHeaders generates block headers, filter header and hashes, and
+// checkpoints from the given genesis. The onCheckpoint method will be called
+// with the current cf header on each checkpoint to modify the derivation of
+// the next interval
+func generateHeaders(genesisBlockHeader *wire.BlockHeader,
+	genesisFilterHeader *chainhash.Hash,
+	onCheckpoint func(*chainhash.Hash)) (*headers, error) {
+
+	var blockHeaders []headerfs.BlockHeader
+	blockHeaders = append(blockHeaders, headerfs.BlockHeader{
+		BlockHeader: genesisBlockHeader,
+		Height:      0,
+	})
+
+	var cfHeaders []headerfs.FilterHeader
+	cfHeaders = append(cfHeaders, headerfs.FilterHeader{
+		HeaderHash: genesisBlockHeader.BlockHash(),
+		FilterHash: *genesisFilterHeader,
+		Height:     0,
+	})
+
+	// The filter hashes (not the filter headers!) will be sent as
+	// part of the CFHeaders response, so we also keep track of
+	// them.
+	genesisFilter, err := builder.BuildBasicFilter(
+		chaincfg.SimNetParams.GenesisBlock, nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to build genesis filter: %v",
+			err)
+	}
+
+	genesisFilterHash, err := builder.GetFilterHash(genesisFilter)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get genesis filter hash: %v",
+			err)
+	}
+
+	var filterHashes []chainhash.Hash
+	filterHashes = append(filterHashes, genesisFilterHash)
+
+	// Also keep track of the current filter header. We use this to
+	// calculate the next filter header, as it commits to the
+	// previous.
+	currentCFHeader := *genesisFilterHeader
+
+	// checkpoints will be the checkpoints passed to
+	// getCheckpointedCFHeaders.
+	var checkpoints []*chainhash.Hash
+
+	for height := uint32(1); height <= maxHeight; height++ {
+		header := heightToHeader(height)
+		blockHeader := headerfs.BlockHeader{
+			BlockHeader: header,
+			Height:      height,
+		}
+
+		blockHeaders = append(blockHeaders, blockHeader)
+
+		// It doesn't really matter what filter the filter
+		// header commit to, so just use the height as a nonce
+		// for the filters.
+		filterHash := chainhash.Hash{}
+		binary.BigEndian.PutUint32(filterHash[:], height)
+		filterHashes = append(filterHashes, filterHash)
+
+		// Calculate the current filter header, and add to our
+		// slice.
+		currentCFHeader = chainhash.DoubleHashH(
+			append(filterHash[:], currentCFHeader[:]...),
+		)
+		cfHeaders = append(cfHeaders, headerfs.FilterHeader{
+			HeaderHash: header.BlockHash(),
+			FilterHash: currentCFHeader,
+			Height:     height,
+		})
+
+		// Each interval we must record a checkpoint.
+		if height%wire.CFCheckptInterval == 0 {
+			// We must make a copy of the current header to
+			// avoid mutation.
+			cfh := currentCFHeader
+			checkpoints = append(checkpoints, &cfh)
+
+			if onCheckpoint != nil {
+				onCheckpoint(&currentCFHeader)
+			}
+		}
+	}
+
+	return &headers{
+		blockHeaders: blockHeaders,
+		cfHeaders:    cfHeaders,
+		checkpoints:  checkpoints,
+		filterHashes: filterHashes,
+	}, nil
+}
+
+// generateResponses generates the MsgCFHeaders messages from the given queries
+// and headers.
+func generateResponses(msgs []wire.Message,
+	headers *headers) ([]*wire.MsgCFHeaders, error) {
+
+	// Craft a response for each message.
+	var responses []*wire.MsgCFHeaders
+	for _, msg := range msgs {
+		// Only GetCFHeaders expected.
+		q, ok := msg.(*wire.MsgGetCFHeaders)
+		if !ok {
+			return nil, fmt.Errorf("got unexpected message %T",
+				msg)
+		}
+
+		// The start height must be set to a checkpoint height+1.
+		if q.StartHeight%wire.CFCheckptInterval != 1 {
+			return nil, fmt.Errorf("unexpexted start height %v",
+				q.StartHeight)
+		}
+
+		var prevFilterHeader chainhash.Hash
+		switch q.StartHeight {
+
+		// If the start height is 1 the prevFilterHeader is set to the
+		// genesis header.
+		case 1:
+			genesisFilterHeader := headers.cfHeaders[0].FilterHash
+			prevFilterHeader = genesisFilterHeader
+
+		// Otherwise we use one of the created checkpoints.
+		default:
+			j := q.StartHeight/wire.CFCheckptInterval - 1
+			prevFilterHeader = *headers.checkpoints[j]
+		}
+
+		resp := &wire.MsgCFHeaders{
+			FilterType:       q.FilterType,
+			StopHash:         q.StopHash,
+			PrevFilterHeader: prevFilterHeader,
+		}
+
+		// Keep adding filter hashes until we reach the stop hash.
+		for h := q.StartHeight; ; h++ {
+			resp.FilterHashes = append(
+				resp.FilterHashes, &headers.filterHashes[h],
+			)
+
+			blockHash := headers.blockHeaders[h].BlockHash()
+			if blockHash == q.StopHash {
+				break
+			}
+		}
+
+		responses = append(responses, resp)
+	}
+
+	return responses, nil
+}
 
 // TestBlockManagerInitialInterval tests that the block manager is able to
 // handle checkpointed filter header query responses in out of order, and when
@@ -52,34 +281,11 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 
 	for _, test := range testCases {
 
-		// Set up the block and filter header stores.
-		tempDir, err := ioutil.TempDir("", "neutrino")
+		bm, hdrStore, cfStore, cleanUp, err := setupBlockManager()
 		if err != nil {
-			t.Fatalf("Failed to create temporary directory: %s",
-				err)
+			t.Fatalf("unable to set up ChainService: %v", err)
 		}
-		defer os.RemoveAll(tempDir)
-
-		db, err := walletdb.Create("bdb", tempDir+"/weks.db")
-		if err != nil {
-			t.Fatalf("Error opening DB: %s", err)
-		}
-		defer db.Close()
-
-		hdrStore, err := headerfs.NewBlockHeaderStore(
-			tempDir, db, &chaincfg.SimNetParams,
-		)
-		if err != nil {
-			t.Fatalf("Error creating block header store: %s", err)
-		}
-
-		cfStore, err := headerfs.NewFilterHeaderStore(
-			tempDir, db, headerfs.RegularFilter,
-			&chaincfg.SimNetParams,
-		)
-		if err != nil {
-			t.Fatalf("Error creating filter header store: %s", err)
-		}
+		defer cleanUp()
 
 		// Keep track of the filter headers and block headers. Since
 		// the genesis headers are written automatically when the store
@@ -89,92 +295,20 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		var blockHeaders []headerfs.BlockHeader
-		blockHeaders = append(blockHeaders, headerfs.BlockHeader{
-			BlockHeader: genesisBlockHeader,
-			Height:      0,
-		})
-
 		genesisFilterHeader, _, err := cfStore.ChainTip()
 		if err != nil {
 			t.Fatal(err)
 		}
 
-		var cfHeaders []headerfs.FilterHeader
-		cfHeaders = append(cfHeaders, headerfs.FilterHeader{
-			HeaderHash: genesisBlockHeader.BlockHash(),
-			FilterHash: *genesisFilterHeader,
-			Height:     0,
-		})
-
-		// The filter hashes (not the filter headers!) will be sent as
-		// part of the CFHeaders response, so we also keep track of
-		// them
-		genesisFilter, err := builder.BuildBasicFilter(
-			chaincfg.SimNetParams.GenesisBlock, nil,
-		)
+		headers, err := generateHeaders(genesisBlockHeader,
+			genesisFilterHeader, nil)
 		if err != nil {
-			t.Fatalf("unable to build genesis filter: %v", err)
-		}
-
-		genesisFilterHash, err := builder.GetFilterHash(genesisFilter)
-		if err != nil {
-			t.Fatal("fail")
-		}
-
-		var filterHashes []chainhash.Hash
-		filterHashes = append(filterHashes, genesisFilterHash)
-
-		// Also keep track of the current filter header. We use this to
-		// calculate the next filter header, as it commits to the
-		// previous.
-		currentCFHeader := *genesisFilterHeader
-
-		// checkpoints will be the checkpoints passed to
-		// getCheckpointedCFHeaders.
-		var checkpoints []*chainhash.Hash
-
-		maxHeight := 20 * uint32(wire.CFCheckptInterval)
-		for height := uint32(1); height <= maxHeight; height++ {
-			header := heightToHeader(height)
-			blockHeader := headerfs.BlockHeader{
-				BlockHeader: header,
-				Height:      height,
-			}
-
-			blockHeaders = append(blockHeaders, blockHeader)
-
-			// It doesn't really matter what filter the filter
-			// header commit to, so just use the height as a nonce
-			// for the filters.
-			filterHash := chainhash.Hash{}
-			binary.BigEndian.PutUint32(filterHash[:], height)
-			filterHashes = append(filterHashes, filterHash)
-
-			// Calculate the current filter header, and add to our
-			// slice.
-			currentCFHeader = chainhash.DoubleHashH(
-				append(filterHash[:], currentCFHeader[:]...),
-			)
-			cfHeaders = append(cfHeaders, headerfs.FilterHeader{
-				HeaderHash: header.BlockHash(),
-				FilterHash: currentCFHeader,
-				Height:     height,
-			})
-
-			// Each interval we must record a checkpoint.
-			if height > 0 && height%wire.CFCheckptInterval == 0 {
-				// We must make a copy of the current header to
-				// avoid mutation.
-				cfh := currentCFHeader
-				checkpoints = append(checkpoints, &cfh)
-			}
-
+			t.Fatalf("unable to generate headers: %v", err)
 		}
 
 		// Write all block headers but the genesis, since it is already
 		// in the store.
-		if err = hdrStore.WriteHeaders(blockHeaders[1:]...); err != nil {
+		if err = hdrStore.WriteHeaders(headers.blockHeaders[1:]...); err != nil {
 			t.Fatalf("Error writing batch of headers: %s", err)
 		}
 
@@ -182,7 +316,7 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 		// written to the store by writing 1/3 of the first interval.
 		if test.partialInterval {
 			err = cfStore.WriteHeaders(
-				cfHeaders[1 : wire.CFCheckptInterval/3]...,
+				headers.cfHeaders[1 : wire.CFCheckptInterval/3]...,
 			)
 			if err != nil {
 				t.Fatalf("Error writing batch of headers: %s",
@@ -190,67 +324,17 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 			}
 		}
 
-		// Set up a chain service with a custom query batch method.
-		cs := &ChainService{
-			BlockHeaders: hdrStore,
-		}
-		cs.queryBatch = func(msgs []wire.Message, f func(*ServerPeer,
-			wire.Message, wire.Message) bool, q <-chan struct{},
-			qo ...QueryOption) {
+		// We set up a custom query batch method for this test, as we
+		// will use this to feed the blockmanager with our crafted
+		// responses.
+		bm.server.queryBatch = func(msgs []wire.Message,
+			f func(*ServerPeer, wire.Message, wire.Message) bool,
+			q <-chan struct{}, qo ...QueryOption) {
 
-			// Craft response for each message.
-			var responses []*wire.MsgCFHeaders
-			for _, msg := range msgs {
-				// Only GetCFHeaders expected.
-				q, ok := msg.(*wire.MsgGetCFHeaders)
-				if !ok {
-					t.Fatalf("got unexpected message %T",
-						msg)
-				}
-
-				// The start height must be set to a checkpoint
-				// height+1.
-				if q.StartHeight%wire.CFCheckptInterval != 1 {
-					t.Fatalf("unexpexted start height %v",
-						q.StartHeight)
-				}
-
-				var prevFilterHeader chainhash.Hash
-				switch q.StartHeight {
-
-				// If the start height is 1 the
-				// prevFilterHeader is set to the genesis
-				// header.
-				case 1:
-					prevFilterHeader = *genesisFilterHeader
-
-				// Otherwise we use one of the created
-				// checkpoints.
-				default:
-					j := q.StartHeight/wire.CFCheckptInterval - 1
-					prevFilterHeader = *checkpoints[j]
-				}
-
-				resp := &wire.MsgCFHeaders{
-					FilterType:       q.FilterType,
-					StopHash:         q.StopHash,
-					PrevFilterHeader: prevFilterHeader,
-				}
-
-				// Keep adding filter hashes until we reach the
-				// stop hash.
-				for h := q.StartHeight; ; h++ {
-					resp.FilterHashes = append(
-						resp.FilterHashes, &filterHashes[h],
-					)
-
-					blockHash := blockHeaders[h].BlockHash()
-					if blockHash == q.StopHash {
-						break
-					}
-				}
-
-				responses = append(responses, resp)
+			responses, err := generateResponses(msgs, headers)
+			if err != nil {
+				t.Fatalf("unable to generate responses: %v",
+					err)
 			}
 
 			// We permute the response order if the test signals
@@ -279,23 +363,10 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 			}
 		}
 
-		// Set up a blockManager with the chain service we defined...
-		bm := blockManager{
-			server: cs,
-			blkHeaderProgressLogger: newBlockProgressLogger(
-				"Processed", "block", log,
-			),
-			fltrHeaderProgessLogger: newBlockProgressLogger(
-				"Verified", "filter header", log,
-			),
-		}
-		bm.newHeadersSignal = sync.NewCond(&bm.newHeadersMtx)
-		bm.newFilterHeadersSignal = sync.NewCond(&bm.newFilterHeadersMtx)
-
-		// ...and call the get checkpointed cf headers method with the
-		// checkpoints we created.
+		// Call the get checkpointed cf headers method with the
+		// checkpoints we created to start the test.
 		bm.getCheckpointedCFHeaders(
-			checkpoints, cfStore, wire.GCSFilterRegular,
+			headers.checkpoints, cfStore, wire.GCSFilterRegular,
 		)
 
 		// Finally make sure the filter header tip is what we expect.
@@ -309,7 +380,7 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 				maxHeight, tipHeight)
 		}
 
-		lastCheckpoint := checkpoints[len(checkpoints)-1]
+		lastCheckpoint := headers.checkpoints[len(headers.checkpoints)-1]
 		if *tip != *lastCheckpoint {
 			t.Fatalf("expected tip to be %v, was %v",
 				lastCheckpoint, tip)

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -411,6 +411,10 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 		// filters from a wrong genesis.
 		wrongGenesis bool
 
+		// intervalMisaligned indicates whether each interval prev hash
+		// should not line up with the previous checkpoint.
+		intervalMisaligned bool
+
 		// partialInterval indicates whether we should write parts of
 		// the first checkpoint interval to the filter header store
 		// before starting the test.
@@ -437,6 +441,21 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 			wrongGenesis:    true,
 			partialInterval: true,
 			firstInvalid:    0,
+		},
+
+		// With intervals not lining up, the second interval should
+		// be determined invalid.
+		{
+			intervalMisaligned: true,
+			firstInvalid:       1,
+		},
+
+		// With misaligned intervals and a partial interval written, the
+		// second interval should be considered invalid.
+		{
+			intervalMisaligned: true,
+			partialInterval:    true,
+			firstInvalid:       1,
 		},
 	}
 
@@ -467,7 +486,16 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 		}
 
 		headers, err := generateHeaders(genesisBlockHeader,
-			genesisFilterHeader, nil)
+			genesisFilterHeader,
+			func(currentCFHeader *chainhash.Hash) {
+				// If we are testing that each interval doesn't
+				// line up properly with the previous, we flip
+				// a bit in the current header before
+				// calculating the next interval checkpoint.
+				if test.intervalMisaligned {
+					currentCFHeader[0] ^= 1
+				}
+			})
 		if err != nil {
 			t.Fatalf("unable to generate headers: %v", err)
 		}
@@ -498,6 +526,20 @@ func TestBlockManagerInvalidInterval(t *testing.T) {
 			if err != nil {
 				t.Fatalf("unable to generate responses: %v",
 					err)
+			}
+
+			// Since we used the generated checkpoints when
+			// creating the responses, we must flip the
+			// PrevFilterHeader bit back before sending them if we
+			// are checking for misaligned intervals. This to
+			// ensure we don't hit the invalid prev hash case.
+			if test.intervalMisaligned {
+				for i := range responses {
+					if i == 0 {
+						continue
+					}
+					responses[i].PrevFilterHeader[0] ^= 1
+				}
 			}
 
 			// Check that the success of the callback match what we

--- a/blockmanager_test.go
+++ b/blockmanager_test.go
@@ -280,6 +280,8 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 	}
 
 	for _, test := range testCases {
+		testDesc := fmt.Sprintf("permute=%v, partial=%v, repeat=%v",
+			test.permute, test.partialInterval, test.repeat)
 
 		bm, hdrStore, cfStore, cleanUp, err := setupBlockManager()
 		if err != nil {
@@ -345,8 +347,17 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 				if test.permute {
 					index = v
 				}
+
+				// Before sending we take a copy of the
+				// message, as we cannot guarantee that it
+				// won't be modified.
+				r := *responses[index]
+
+				// Let the blockmanager handle the message.
 				if !f(nil, msgs[index], responses[index]) {
-					t.Fatalf("got response false")
+					t.Fatalf("got response false on "+
+						"send of index %d: %v",
+						index, testDesc)
 				}
 
 				// If we are not testing repeated responses, go
@@ -356,8 +367,10 @@ func TestBlockManagerInitialInterval(t *testing.T) {
 				}
 
 				// Otherwise resend the response we just sent.
-				if !f(nil, msgs[index], responses[index]) {
-					t.Fatalf("got response false")
+				if !f(nil, msgs[index], &r) {
+					t.Fatalf("got response false on "+
+						"resend of index %d: %v",
+						index, testDesc)
 				}
 
 			}

